File: server/src/main/com/broll/gainea/server/core/objects/buffs/BuffableInt.java
package com.broll.gainea.server.core.objects.buffs;


public class BuffableInt<T> extends BuffableValue<T, IntBuff, Integer> {

    private int minValue = Integer.MIN_VALUE;
    private int maxValue = Integer.MAX_VALUE;

    public BuffableInt(T object, int value) {
        super(object);
        this.value = value;
        recalc();
    }

    public BuffableInt copy(T object) {
        BuffableInt copy = new BuffableInt(object, value);
        copy.minValue = this.minValue;
        copy.maxValue = this.maxValue;
        copy.buffs = this.buffs;
        return copy;
    }

    protected void recalc() {
        buffedValue = value.intValue();
        for (IntBuff buff : buffs) {
            int modifier = buff.getModifier();
            switch (buff.getType()) {
                case ADD:
                    buffedValue += modifier;
                    break;
                case MULTIPLY:
                    buffedValue *= modifier;
                    break;
                case SET:
                    buffedValue = modifier;
                    break;
            }
        }
        if (buffedValue < minValue) {
            buffedValue = minValue;
        } else if (buffedValue > maxValue) {
            buffedValue = maxValue;
        }
    }

    public void setMinValue(int minValue) {
        this.minValue = minValue;
        recalc();
    }

    public void setMaxValue(int maxValue) {
        this.maxValue = maxValue;
        recalc();
    }

    public void addValue(int value) {
        this.value += value;
        recalc();
    }


}

File: server/src/main/com/broll/gainea/server/core/objects/buffs/BuffableInt.kt
package com.broll.gainea.server.core.objects.buffs

class BuffableInt<T>(target: T, value: Int) : BuffableValue<T, Int, IntBuff>(target, value) {
    private var minValue = Int.MIN_VALUE
    private var maxValue = Int.MAX_VALUE

    init {
        this.value = value
        recalc()
    }

    fun copy(target: T): BuffableInt<T> {
        val copy = BuffableInt(target, this.rootValue)
        copy.minValue = minValue
        copy.maxValue = maxValue
        copy.buffs = buffs
        copy.recalc()
        return copy
    }

    fun frozenCopy(target: T): BuffableInt<T> = copy(target).also { it.freeze() }

    override fun recalc() {
        buffedValue = rootValue
        for (buff in buffs) {
            val modifier = buff.modifier
            buffedValue = when (buff.type) {
                BuffType.ADD -> buffedValue + modifier
                BuffType.MULTIPLY -> buffedValue * modifier
                BuffType.SET -> modifier
            }
        }
        if (buffedValue < minValue) {
            buffedValue = minValue
        } else if (buffedValue > maxValue) {
            buffedValue = maxValue
        }
    }

    fun setMinValue(minValue: Int) {
        this.minValue = minValue
        recalc()
    }

    fun setMaxValue(maxValue: Int) {
        this.maxValue = maxValue
        recalc()
    }

    fun addValue(value: Int) {
        this.rootValue += value
        recalc()
    }
}
